### PR TITLE
drivers: i3c: cdns: add ddr, ibi tir, direct ccc def byte support

### PR DESCRIPTION
**Pull request by XenuIsWatching**
Revision 1p7 of the silicon added support for IBI TIR, full word fifo while target usage, defining byte with direct CCCs. This also adds support of HDR-DDR. This doesn't contain support for the new flow nack control while acting as a target

This also fixes short ccc commands where multiple targets in a payload for the same ccc.

Modification of the i3c_target_tx_write had to be done in order to distinguish the difference for each fifo to be written to.

### CODE REVIEW

**Comment by XenuIsWatching** [comment]
I'm wondering if it may be use to have a `hdr-ddr.h` within `zephyr/include/zephyr/drivers/i3c` which will contain the same i3c helper functions such as `i3c_write` but one that sets up the flags for a DDR write.

Some of the Macros that I'm using here such as the preamble defines and CRC token also probably could be defined in this header.

```c
static inline int i3c_ddr_write(struct i3c_device_desc *target,
			    const uint8_t *buf, uint32_t num_bytes)
{
	struct i3c_msg msg;

	msg.buf = (uint8_t *)buf;
	msg.len = num_bytes;
	msg.flags = I3C_MSG_WRITE | I3C_MSG_STOP |  I3C_MSG_HDR;
	msg.hdr_mode = I3C_MSG_HDR_DDR;

	return i3c_transfer(target, &msg, 1);
}
```

**Review by dcpleung — DISMISSED** [review]

**Comment by dcpleung** [comment]
> I'm wondering if it may be useful to have a `hdr-ddr.h` within `zephyr/include/zephyr/drivers/i3c` which will contain the same i3c helper functions such as `i3c_write` but one that sets up the flags for a DDR write.
> 
> Some of the Macros that I'm using here such as the preamble defines and CRC token also probably could be defined in this header.
> 
> Also the even and odd parity, the parity even read adjuster bit, and crc5 function are also something that could be put something more "common"

Sounds like a good idea as these can be defined as inline function which (kind of) act like a wrapper.

**Comment by alvsun** [comment]
Hi, I'm currently working on DDR mode. 
The command code in HDR-DDR command word format provides 128 possible read (0x00 - 0x7F) and write (0x80 - 0xFF) commands for users. 
Would it be more flexible to add this argument to the struct i3c_msg and update the helper functions as follows?

```c
struct i3c_msg {
      ......

	/**
	 * HDR command code field
	 *
	 * For DDR,
	 * Write: 8’h00 to 8’h7F
	 * Read : 8’h80 to 8’hFF
	 */
	uint8_t			hdr_cmd_code;
};
```

```c
static inline int i3c_ddr_write(struct i3c_device_desc *target, uint8_t ddr_cmd,
				uint8_t *buf, uint32_t num_bytes)
{
	struct i3c_msg msg;

	msg.buf = buf;
	msg.len = num_bytes;
	msg.flags = I3C_MSG_WRITE | I3C_MSG_STOP | I3C_MSG_HDR;
	msg.hdr_mode = I3C_MSG_HDR_DDR;
	msg.hdr_cmd_code = ddr_cmd;

	return i3c_transfer(target, &msg, 1);
}
```

**Comment by XenuIsWatching** [comment]
> Hi, I'm currently working on DDR mode. The command code in HDR-DDR command word format provides 128 possible read (0x00 - 0x7F) and write (0x80 - 0xFF) commands for users. Would it be more flexible to add this argument to the `struct i3c_msg` and update the helper functions as follows?
> 
> ```c
> struct i3c_msg {
>       ......
> 
> 	/**
> 	 * HDR command code field
> 	 *
> 	 * For DDR,
> 	 * Write: 8’h00 to 8’h7F
> 	 * Read : 8’h80 to 8’hFF
> 	 */
> 	uint8_t			hdr_cmd_code;
> };
> ```
> 
> ```c
> static inline int i3c_ddr_write(struct i3c_device_desc *target, uint8_t ddr_cmd,
> 				uint8_t *buf, uint32_t num_bytes)
> {
> 	struct i3c_msg msg;
> 
> 	msg.buf = buf;
> 	msg.len = num_bytes;
> 	msg.flags = I3C_MSG_WRITE | I3C_MSG_STOP | I3C_MSG_HDR;
> 	msg.hdr_mode = I3C_MSG_HDR_DDR;
> 	msg.hdr_cmd_code = ddr_cmd;
> 
> 	return i3c_transfer(target, &msg, 1);
> }
> ```

yeah... that is something that needs to be considered as it is apart of the specification.  I think the same applies to the TSL and TSP HDR modes as well. That command code may also just want only be the 7 bits only with that read/write bit being implied by the I3C_MSG_WRITE flag.

I think that would be something of good value... but probably in separate PR along with the "static inline" ddr helper functions. Is this something you can work on?

**Comment by XenuIsWatching** [comment]
I need to come back to with this and add support for using the HDR 7b command now that https://github.com/zephyrproject-rtos/zephyr/pull/71170 has landed. Should be a fairly simple change

**Comment by XenuIsWatching** [comment]
> I need to come back to with this and add support for using the HDR 7b command now that #71170 has landed. Should be a fairly simple change

It's been added and it's now fully really for review (again)

**Review by dcpleung — DISMISSED** [review]

File: include/zephyr/drivers/i3c.h [hunk 1]
```diff
@@ -864,11 +864,12 @@ __subsystem struct i3c_driver_api {
 	 * @param dev Pointer to the controller device driver instance.
 	 * @param buf Pointer to the buffer
 	 * @param len Length of the buffer
+	 * @param hdr_mode HDR mode of the buffer
```
Comment (kartben) [inline include/zephyr/drivers/i3c.h:867]: Should it really say "... of the buffer"?

File: include/zephyr/drivers/i3c/target_device.h [hunk 2]
```diff
@@ -233,14 +233,15 @@ __subsystem struct i3c_target_driver_api {
  *            driver configured in target mode.
  * @param buf Pointer to the buffer
  * @param len Length of the buffer
+ * @param hdr_mode HDR mode of the buffer see @c I3C_MSG_HDR_MODE*
```
Comment (kartben) [inline include/zephyr/drivers/i3c/target_device.h:236]: Ditto

File: include/zephyr/drivers/i3c.h
Comment (kartben) [inline include/zephyr/drivers/i3c.h:1]: Can you put the formatting changes in a separate commit?

**Review by kartben — DISMISSED** [review]

File: include/zephyr/drivers/i3c.h (hunk 1 above)
Comment (XenuIsWatching) [reply include/zephyr/drivers/i3c.h:867]: yeah, you're probably right, will remove

File: include/zephyr/drivers/i3c/target_device.h (hunk 2 above)
Comment (XenuIsWatching) [reply include/zephyr/drivers/i3c/target_device.h:236]: yep, removed

File: include/zephyr/drivers/i3c.h
Comment (XenuIsWatching) [reply include/zephyr/drivers/i3c.h:1]: What formatting changes?

Comment (kartben) [reply include/zephyr/drivers/i3c.h:1]: Sorry looks like I misclicked on my phone: in the .c file, not this one :)

Comment (XenuIsWatching) [reply include/zephyr/drivers/i3c.h:1]: oh, that may be too much of a pain at this point :( Actually alot of those formating changes came from adding new defines and things had to realigned.

**Review by dcpleung — APPROVED** [review]

**Review by gmarull — APPROVED** [review]
